src/tools/rbd_mirror/InstanceReplayer.cc

```diff
@@ -541,6 +638,225 @@ void InstanceReplayer<I>::schedule_image_state_check_task() {
   m_threads->timer->add_event_after(after, m_image_state_check_task);
 }
 
+template <typename I>
+void InstanceReplayer<I>::start_group_replayer(
+    GroupReplayer<I> *group_replayer) {
+  dout(10) << dendl;
+  ceph_assert(ceph_mutex_is_locked(m_lock));
+
+  std::string global_group_id = group_replayer->get_global_group_id();
+  if (!group_replayer->is_stopped()) {
+    if (group_replayer->needs_restart()) {
+      stop_group_replayer(group_replayer, new C_TrackedOp(m_async_op_tracker,
+                                                          nullptr));
+    } else {
+      group_replayer->sync_group_names();
+    }
+    return;
+  } else if (group_replayer->is_blocklisted()) {
+    derr << "global_group_id=" << global_group_id << ": blocklisted detected "
+         << "during group replay" << dendl;
+    m_blocklisted = true;
+    return;
+  } else if (group_replayer->is_finished()) {
+    // TODO temporary until policy integrated
+    dout(5) << "removing group replayer for global_group_id="
+            << global_group_id << dendl;
+    m_group_replayers.erase(group_replayer->get_global_group_id());
+    group_replayer->destroy();
+    return;
+  } else if (m_manual_stop) {
+    return;
+  }
+
+  dout(10) << "global_group_id=" << global_group_id << dendl;
+  group_replayer->start(new C_TrackedOp(m_async_op_tracker, nullptr), false);
+}
+
+template <typename I>
+void InstanceReplayer<I>::queue_start_group_replayers() {
+  dout(10) << dendl;
+
+  Context *ctx = create_context_callback<
+    InstanceReplayer, &InstanceReplayer<I>::start_group_replayers>(this);
+  m_async_op_tracker.start_op();
+  m_threads->work_queue->queue(ctx, 0);
+}
+
+template <typename I>
+void InstanceReplayer<I>::start_group_replayers(int r) {
+  dout(10) << dendl;
+
+  std::lock_guard locker{m_lock};
+  if (m_on_shut_down != nullptr) {
+    m_async_op_tracker.finish_op();
+    return;
+  }
+
+  uint64_t group_count = 0;
+  uint64_t warning_count = 0;
+  uint64_t error_count = 0;
+  for (auto it = m_group_replayers.begin();
+       it != m_group_replayers.end();) {
+    auto current_it(it);
+    ++it;
+
+    ++group_count;
+    auto health_state = current_it->second->get_health_state();
+    if (health_state == image_replayer::HEALTH_STATE_WARNING) {
+      ++warning_count;
+    } else if (health_state == image_replayer::HEALTH_STATE_ERROR) {
+      ++error_count;
+    }
+
+    start_group_replayer(current_it->second);
+  }
+
+  m_service_daemon->add_or_update_namespace_attribute(
+    m_local_io_ctx.get_id(), m_local_io_ctx.get_namespace(),
+    SERVICE_DAEMON_ASSIGNED_COUNT_KEY, group_count);
```

Comment:
This looks like it will need to include the image error counts as well - it looks like the group and image replayers will overwrite each others values.